hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java

```diff
@@ -901,6 +919,53 @@ public static HoodieData<HoodieRecord> convertMetadataToRecordIndexRecords(Hoodi
     }
   }
 
+  static Set<String> getValidRecordKeysForFileSlice(HoodieTableMetaClient metaClient,
```

Comment:
Does the file group reading now add additional latency compared to before?  Should we consider optimizations for `EVENT_TIME_ORDERING` that can avoid such merging?  Also is the behavior consistent with global index, i.e., once the record is deleted through a log file in MOR table, the record no longer belongs to the file group even though the record exists in the base file?

Reply:
>  to add ordering value to RLI entry of each record, so that we can use the order value when updating the metadata payload record in RLI in the correct ordering based on the merge mode.

This looks like the right fix for me. And it is very related to the delete block processing time semantics lost issue:
https://github.com/apache/hudi/pull/12971#discussion_r1993030977